### DISCUSSION TITLE
Prevent griefing

### DISCUSSION THREAD

**Discussion by npabisz** [OpenTTD/OpenTTD#11793] [Features]
Right now its very easy for a griefer to ruin the experience of other players in multiplayer games. We should try to do something about it.

**IP bans**

Nowadays almost everyone has dynamic IP. We should have better identification of users which they cannot alter easily, eg. fingerprint calculated base on hardware, mac address, etc. It will still be possible to bypass, but much harder that reconnecting ISP to get new IP address.

**Game Scripts**

Game scripts should have more control over actions taken by player. Eg. prevent them from demolishing, rising/lowering tiles, building.

**Comment by 2TallTyler** [comment]
How would a Game Script determine which players are allowed to perform which actions? Or do you mean that server administrators should be able to prohibit some actions for all players?

**Reply by npabisz** [reply]
@2TallTyler I was think about some kind of events and event listeners which could respond to these events. Something like eg. javascript which can listen to clicks on buttons and prevent events from being fired or propagated. Then you could set listeners for specific actions in script and game would know that it needs first to ask GS for response.

**Reply by PeterN** [reply]
GS isn't event based and cannot be interrupted or respond in real time, but also is meant for implementing gameplay behaviour, not anti-griefing rules.

**Reply by npabisz** [reply]
@PeterN I know its not event based right now, however this could bring many new posibilities. Also preventing from building in certain situations it is gameplay behaviour

**Comment by rubidium42** [comment]
Given OpenTTD is open source, the griefer will have easily access to the code that generates the fingerprint. If someone want to grief, then recompiling the game so it always generate a different fingerprint is simple enough.

I actually think that you, as griefer, have less control over the IP address from your ISP, than over things like your MAC address. Just Google for changing your MAC address and you'll quickly find at least a couple of ways to do it.
If I want a different IP from my ISP, I first have to let the DHCP lease expire. Given the least is for a week, just reconnecting will likely not yield a different IP address for me.

**Reply by npabisz** [reply]
@rubidium42 I'm talking about casual players. If someone is able to read, understand code and make changes just in order to grief then at this point it doesn't matter what way of preventing grief the game has. But how many people like that would make such effort to bypass ban in game? I belive the game called Apex Legends was pretty successfully at hardware bans and we are talking about the game which had millions of players and cheaters.

**Comment by andythenorth** [comment]
"I only play with friends, and people I trust" :)

**Comment by TrueBrain** [comment]
The issue with a discussion that starts with a few concrete examples, that it is easy to get lost in those examples, and not think any further. So I am just going to read: "what could we do to prevent griefing further", and take it as that :)

IP-bans are still the most efficient, although it is after-the-fact. People are know how to avoid an IP-ban are often also creative enough to avoid other types of ban, so I doubt there is really much value in doing anything more on that level. And, to voice an opinion on that matter, fingerprinting a machine is something that really doesn't belong in OpenTTD, and smells like spyware, tracking, etc. Let's please not do that because a select few cannot behave.

GameScript is designed to add gaming elements to the game. Not to intercept every command and judge whether that is good or not. Just because we have a hammer, let's not make everything a nail. "If Your Only Tool Is a Hammer Then Every Problem Looks Like a Nail".
But basically what you ask for is some kind of [Zero Trust](https://www.microsoft.com/en-us/security/business/zero-trust) model, where every action is being evaluated whether it should be allowed or not. This could be done in a new interface, specially designed for this. Lot of work, for a relatively small market; so finding someone willing to build that, can be challenging.

Another approach that has been talked about a few times, but nobody actually implemented, is authentication. Where a server can say: I only allow clients that are authenticated against this authentication server. And users first need to make an account there, before they can join. This is especially useful in a cluster of servers, where one owner runs multiple servers. Or rank-servers, where they track how well you did over the last N games. As now you can track an individual easier.
A good anti-grieving in such scenario would be "social score", where if people behave bad, they lose score, and can no longer join certain servers. And slowly they end up on a server that has other grievers, and they can enjoy each other's company.
Of course the authentication server needs to prevent people from creating new accounts when ever they are banned / have a low score, but that is mostly a solved problem already.

But, it all depends on people actually donating their time and effort in building these kind of things. Ideas we have plenty, people to execute it not so much. And all ideas here are not simple changes that are done in an hour or so. That in reality is the actual problem here. Time.

**Reply by npabisz** [reply]
@TrueBrain I was about to comment that these were just my ideas and we can talk about other ideas. I like the idea of the authentication server and social score. However I don't know if having different accounts on each authentication servers are good idea. This could work if there was one global authentication and we could take care of social score internally having user id. Not sure how steam is handled now, but maybe this is a good time to think about authentication

**Comment by DefinitelyNotRau117** [comment]
Another solution would be to limit destructive actions by default.

For example, change build_object_frame_burst to 64 for all existing cfg, where this setting has not been changed. This will make it more difficult to buy large fields of land - the most annoying type of griefing. For a normal game, 8x8 should be enough, and if the player needs to lay more, they will have to click several times. A small price to make griefing more difficult.

Now this setting has a quite large value of 2048 ~ 45x45 and is even more hidden in openttd.cfg. 
_To avoid discomfort from the limit in a single game, it is worth dividing this setting into 2 - for single and server games._

**Comment by LC-Zorg** [comment]
Some or even most of the abuses could be significantly reduced or even eliminated by adding appropriate features to the game. I think in some cases it can be relatively simple.

Examples of some common abuses and solutions:

**1. Deliberate or resulting from inattention destruction of other companies vehicles at level crossings**
S: Adding a function with a setting that prevents building tracks on other companies roads.

**2. Blocking ship routes**
S: Adding a function that builds invisible buoys between the points of the route along which ships move. The buoys will disappear on their own if no ship is moving along the designated route. As long as at least one ship uses this route (is in motion), invisible buoys are unremovable and therefore it is impossible to block the route either by another player or [by the growing city](https://github.com/OpenTTD/OpenTTD/discussions/11735).
Invisible buoys can be visible as a shipping line on the water (something everyone may recognize from real maps)
![Shipping Lanes - technical and gameplay view](https://github.com/OpenTTD/OpenTTD/assets/67662046/b1764af0-cfcf-49ce-b291-ea387e730609)

**3. Stealing products or raw materials**
S: Adding a function with a setting that will limit the ability to service the enterprise (taking products) only by one company. Only the first one or that company which supplies raw materials to the enterprise would be able to receive the products. It would also be possible to create a function where the number of products that the company can take depends on the amount of delivered raw materials

**4. Uninvited entry into the company without a password**
S: Modification of the nickname selection window and forcing the player to create a password before he setting up a company

**5. Establishing several companies in the game**
S: Adding a function with a setting that prevents such operation from the same computer.

**6. Sinking the map**
S: Adding a function with a setting that prevents the terrain from being excessively lowered to level 0. This could be the same setting as for excessive terraforming.

**7. Draining the map**
S: Adding a function with a setting that prevents excessive removal of water tiles. Here it may also be the same setting as in the case of excessive terraforming.

**8. Removal or cutting rivers**
There are at least two possible solutions:
S1. Adding a function with a setting that completely prevents the removal of river tiles
S2. Adding a function with a setting that prevents cutting the connection between river sources and the sea - the player must first build an alternative route (which will be saved) and only then he can remove the selected part.

**9. Blocking another player's construction / Sabotaging a station**
S: Adding a feature with a setting to prevent building too close to another company. Different distances may be for routes and different for stations. This will not fully prevent such events, but it will limit them.

**10. Buying the entire map**
The problem appeared with the addition of the possibility of purchasing a larger area at one time. Before adding it, I mentioned that this could be a problem in online games. ;) I think the solution here would be to add a setting for the maximum amount of land purchased. Existing limits by frames, even if used, may not be fully effective on servers where the administrator wanted to allow some freedom, but at the same time cannot constantly control the game, and in the case of the current solutions, time is on the troll's side.

**11. Intentional or due to poor knowledge of the game blocking the roads by another player's vehicles**
This can be partially prevented by existing settings prohibiting the construction of stops on the other players or city roads. Another problem is that players using other companies roads often cause traffic jams or significantly slow down traffic. I see two solutions here, the second one being conditional on the introduction of the first one:
S1. Adding a feature with a setting or changing the game to prevent another player from expanding another player's roads to connect them to his own roads.
S2. Adding a function with setting fees for using other players roads or the possibility of building toll gates that players could place on their roads (this would also make investing in the construction of roads one of the ways to earn money in the game, this requires a change from point above)


The features mentioned above would be solutions to prevent abuse. However, when it comes to keeping annoying players away from the game or discourage them from doing so, I have an idea related to the method of authentication, maybe naive, maybe there are other, proven methods, and I think I wrote about it somewhere in a slightly different form, but maybe it will be useful.

I think that authentication could be done by comparing the add-ons the player has, the dates they were downloaded, and the number and dates of saves. The OTTD main server would also record login times, also to download add-ons. The player could have different authentication levels depending on the number of logins (max 1 daily to score). This means that if he changed the verification basis by deleting or moving files and modifying the game, his credibility would be reduced to zero and he would have to build it again. Thus, he would lose the opportunity to play on servers that require an appropriate level of credibility.

**Comment by MartinRe63** [comment]
I was just attacked maybe by a dynamic IP and my company was closed :( . I do not really understand what the issue is. Just do not use the IP, but use the MAC address of the computer. OpenTTD is still a normal software, which has access to the MAC and given that it's somewhat easy to identify if the same client is using 2 companies. And I agree with the message before. The multiplayer game migth be improved with improvements in the game --- but I think developers capacity of the game is limited.

**Reply by telk5093** [reply]
MAC address can be changed easily
